### ISSUE TITLE
`Az.Resources` v6.6.1 fails to import to Azure Automation Account

### ISSUE THREAD

**Issue by o-l-a-v** [Azure/azure-powershell#21647]
See [temp workaround](https://github.com/Azure/azure-powershell/issues/21647#issuecomment-1521820690) if you are affected.

### Description

Tried three times to import `Az.Resources` v6.6.1 to an Azure Automation Account, but it fails.

I imported/updated `Az.Accounts` to v2.12.2 prior to this.

This is using the built in "Browse gallery" method.

![image](https://user-images.githubusercontent.com/6450056/234285492-4e4f41a9-b556-43f5-998a-dd55673c6741.png)

### Issue script & Debug output

```PowerShell
None.
```

### Environment data

```PowerShell
5.1 in Azure Automation Account
```

### Module versions

```PowerShell
Az.Accounts v2.12.2
Az.Resources v6.6.1
```

### Error output

> Error importing the module Az.Resources. Import failed with the following error: Orchestrator.Shared.AsyncModuleImport.ModuleImportException: An error occurred when trying to import the module to an internal PowerShell session. Either the module dependencies are not imported correctly or the module is unsupported. Internal Error Message: The type initializer for 'Microsoft.Azure.Commands.Common.AzModule' threw an exception. The type initializer for 'Microsoft.Azure.Commands.Common.AzModule' threw an exception. . at Orchestrator.Activities.SetModuleVersion.GetPsModuleVersion(String moduleName, String moduleFullPath, Guid accountId, Guid moduleVersionId, InitialSessionState defaultSessionState) at Orchestrator.Activities.SetModuleVersion.InsertModuleVersion(String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String moduleFullyQualifiedName, ActivitiesMetadataWorkflowExtension activitiesMetadataWorkflowExtension, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, ContentUriInfo contentUri, String storageUri, Int32 moduleVersion) at Orchestrator.Activities.SetModuleVersion.ExecuteInternal(CodeActivityContext context, String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String modulePath, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, String storageUri, Int32 moduleVersion) at Orchestrator.Activities.SetModuleVersion.Execute(CodeActivityContext context) at System.Activities.CodeActivity.InternalExecute(ActivityInstance instance, ActivityExecutor executor, BookmarkManager bookmarkManager) at System.Activities.Runtime.ActivityExecutor.ExecuteActivityWorkItem.ExecuteBody(ActivityExecutor executor, BookmarkManager bookmarkManager, Location resultLocation)

**Comment by o-l-a-v** [comment]
More info:

Running locally, `Az.Resources` v6.6.1 fails to import on Windows PowerShell 5.1, but only when using VSCode and the [VSCode PowerShell Extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.PowerShell).

* PowerShell.exe terminal works fine.
* PowerShell ISE works fine
* VSCode with PS extension had no problems with `Az.Accounts` v2.12.1 and `Az.Resources` v6.6.0.

Some assembly conflict?

Output with verbose and debug set to 'Continue'.

<details><summary>Click to view</summary>

```powershell
PS C:\Users\ORB\OneDrive - Nasjonal kommunikasjonsmyndighet\Kode> Import-Module -Name Az.Resources
VERBOSE: Loading module from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Az.Resources.psd1'.
VERBOSE: Cannot verify the Microsoft .NET Framework version 4.7.2 because it is not included in the list of permitted versions.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Microsoft.Azure.Management.Authorization.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Microsoft.Azure.Management.Authorization.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Microsoft.Azure.Management.ResourceManager.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Microsoft.Azure.Management.ResourceManager.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Microsoft.Azure.Management.ManagementGroups.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Microsoft.Azure.Management.ManagementGroups.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Microsoft.Extensions.Caching.Abstractions.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Microsoft.Extensions.Caching.Abstractions.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Microsoft.Extensions.Caching.Memory.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Microsoft.Extensions.Caching.Memory.dll'.
VERBOSE: Loading 'Assembly' from path
'C:\Users\ORB\.vscode\extensions\ms-vscode.powershell-2023.3.3\modules\PowerShellEditorServices\bin\Common\Microsoft.Extensions.DependencyInjection.Abstractions.dll'.        
VERBOSE: Loading 'Assembly' from path
'C:\Users\ORB\.vscode\extensions\ms-vscode.powershell-2023.3.3\modules\PowerShellEditorServices\bin\Common\Microsoft.Extensions.DependencyInjection.Abstractions.dll'.        
VERBOSE: Loading 'Assembly' from path
'C:\Users\ORB\.vscode\extensions\ms-vscode.powershell-2023.3.3\modules\PowerShellEditorServices\bin\Common\Microsoft.Extensions.Options.dll'.
VERBOSE: Loading 'Assembly' from path
'C:\Users\ORB\.vscode\extensions\ms-vscode.powershell-2023.3.3\modules\PowerShellEditorServices\bin\Common\Microsoft.Extensions.Options.dll'.
VERBOSE: Loading 'Assembly' from path
'C:\Users\ORB\.vscode\extensions\ms-vscode.powershell-2023.3.3\modules\PowerShellEditorServices\bin\Common\Microsoft.Extensions.Primitives.dll'.
VERBOSE: Loading 'Assembly' from path
'C:\Users\ORB\.vscode\extensions\ms-vscode.powershell-2023.3.3\modules\PowerShellEditorServices\bin\Common\Microsoft.Extensions.Primitives.dll'.
VERBOSE: Loading 'Assembly' from path
'C:\Users\ORB\.vscode\extensions\ms-vscode.powershell-2023.3.3\modules\PowerShellEditorServices\bin\Common\System.Runtime.CompilerServices.Unsafe.dll'.
VERBOSE: Loading 'Assembly' from path
'C:\Users\ORB\.vscode\extensions\ms-vscode.powershell-2023.3.3\modules\PowerShellEditorServices\bin\Common\System.Runtime.CompilerServices.Unsafe.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\bin\Az.MSGraph.private.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\bin\Az.MSGraph.private.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Authorization.Autorest\bin\Az.Authorization.private.dll'.   
VERBOSE: Loading 'Assembly' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Authorization.Autorest\bin\Az.Authorization.private.dll'.   
VERBOSE: Loading 'FormatsToProcess' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Resources.format.ps1xml'.
VERBOSE: Loading 'FormatsToProcess' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\ResourceManager.format.ps1xml'.
VERBOSE: Loading 'FormatsToProcess' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\ResourceManager.generated.format.ps1xml'.
VERBOSE: Loading 'FormatsToProcess' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Tags.format.ps1xml'.
VERBOSE: Loading 'FormatsToProcess' from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\Az.MSGraph.format.ps1xml'.
VERBOSE: Loading 'FormatsToProcess' from path
'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\Authorization.Autorest\Az.Authorization.format.ps1xml'.
VERBOSE: Loading module from path 'C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\Az.MSGraph.psm1'.
Loaded Module 'Az.Accounts'
VERBOSE: Importing cmdlet 'Export-CmdletSurface'.
VERBOSE: Importing cmdlet 'Export-ExampleStub'.
VERBOSE: Importing cmdlet 'Export-FormatPs1xml'.
VERBOSE: Importing cmdlet 'Export-HelpMarkdown'.
VERBOSE: Importing cmdlet 'Export-ModelSurface'.
VERBOSE: Importing cmdlet 'Export-ProxyCmdlet'.
VERBOSE: Importing cmdlet 'Export-Psd1'.
VERBOSE: Importing cmdlet 'Export-TestStub'.
VERBOSE: Importing cmdlet 'Get-CommonParameter'.
VERBOSE: Importing cmdlet 'Get-ModuleGuid'.
VERBOSE: Importing cmdlet 'Get-ScriptCmdlet'.
VERBOSE: Importing cmdlet 'Add-AzADApplicationKey_Add'.
VERBOSE: Importing cmdlet 'Add-AzADApplicationKey_AddExpanded'.
VERBOSE: Importing cmdlet 'Add-AzADApplicationPassword_Add'.
VERBOSE: Importing cmdlet 'Add-AzADApplicationPassword_AddExpanded'.
VERBOSE: Importing cmdlet 'Add-AzADServicePrincipalKey_Add'.
VERBOSE: Importing cmdlet 'Add-AzADServicePrincipalKey_AddExpanded'.
VERBOSE: Importing cmdlet 'Add-AzADServicePrincipalPassword_Add'.
VERBOSE: Importing cmdlet 'Add-AzADServicePrincipalPassword_AddExpanded'.
VERBOSE: Importing cmdlet 'Get-AzADAppFederatedCredential_Get'.
VERBOSE: Importing cmdlet 'Get-AzADAppFederatedCredential_List'.
VERBOSE: Importing cmdlet 'Get-AzADApplication_Get'.
VERBOSE: Importing cmdlet 'Get-AzADApplication_List'.
VERBOSE: Importing cmdlet 'Get-AzADGroupMember_List'.
VERBOSE: Importing cmdlet 'Get-AzADGroup_Get'.
VERBOSE: Importing cmdlet 'Get-AzADGroup_List'.
VERBOSE: Importing cmdlet 'Get-AzADOrganization_List'.
VERBOSE: Importing cmdlet 'Get-AzADServicePrincipal_Get'.
VERBOSE: Importing cmdlet 'Get-AzADServicePrincipal_List'.
VERBOSE: Importing cmdlet 'Get-AzADUserOwnedApplication_List'.
VERBOSE: Importing cmdlet 'Get-AzADUserOwnedObject_List'.
VERBOSE: Importing cmdlet 'Get-AzADUserSigned_Get'.
VERBOSE: Importing cmdlet 'Get-AzADUser_Get'.
VERBOSE: Importing cmdlet 'Get-AzADUser_List'.
VERBOSE: Importing cmdlet 'New-AzADAppFederatedCredential_CreateExpanded'.
VERBOSE: Importing cmdlet 'New-AzADApplication_CreateExpanded'.
VERBOSE: Importing cmdlet 'New-AzADGroupGraphRefMember_Create'.
VERBOSE: Importing cmdlet 'New-AzADGroupGraphRefMember_CreateExpanded'.
VERBOSE: Importing cmdlet 'New-AzADGroup_CreateExpanded'.
VERBOSE: Importing cmdlet 'New-AzADOrganization_Create'.
VERBOSE: Importing cmdlet 'New-AzADOrganization_CreateExpanded'.
VERBOSE: Importing cmdlet 'New-AzADServicePrincipal_CreateExpanded'.
VERBOSE: Importing cmdlet 'New-AzADUser_CreateExpanded'.
VERBOSE: Importing cmdlet 'Remove-AzADAppFederatedCredential_Delete'.
VERBOSE: Importing cmdlet 'Remove-AzADApplicationKey_Remove'.
VERBOSE: Importing cmdlet 'Remove-AzADApplicationKey_RemoveExpanded'.
VERBOSE: Importing cmdlet 'Remove-AzADApplicationPassword_Remove'.
VERBOSE: Importing cmdlet 'Remove-AzADApplicationPassword_RemoveExpanded'.
VERBOSE: Importing cmdlet 'Remove-AzADApplication_Delete'.
VERBOSE: Importing cmdlet 'Remove-AzADGroupRefMember_Delete'.
VERBOSE: Importing cmdlet 'Remove-AzADGroup_Delete'.
VERBOSE: Importing cmdlet 'Remove-AzADServicePrincipalKey_Remove'.
VERBOSE: Importing cmdlet 'Remove-AzADServicePrincipalKey_RemoveExpanded'.
VERBOSE: Importing cmdlet 'Remove-AzADServicePrincipalPassword_Remove'.
VERBOSE: Importing cmdlet 'Remove-AzADServicePrincipalPassword_RemoveExpanded'.
VERBOSE: Importing cmdlet 'Remove-AzADServicePrincipal_Delete'.
VERBOSE: Importing cmdlet 'Remove-AzADUser_Delete'.
VERBOSE: Importing cmdlet 'Update-AzADAppFederatedCredential_UpdateExpanded'.
VERBOSE: Importing cmdlet 'Update-AzADApplication_UpdateExpanded'.
VERBOSE: Importing cmdlet 'Update-AzADGroup_UpdateExpanded'.
VERBOSE: Importing cmdlet 'Update-AzADServicePrincipal_UpdateExpanded'.
VERBOSE: Importing cmdlet 'Update-AzADUser_UpdateExpanded'.
Register-AzModule : The type initializer for 'Microsoft.Azure.Commands.Common.AzModule' threw an exception.
At C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\Az.MSGraph.psm1:49 char:13
+   $VTable = Register-AzModule
+             ~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [Register-AzModule], TypeInitializationException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Common.RegisterAzModule

PS C:\Users\ORB\OneDrive - Nasjonal kommunikasjonsmyndighet\Kode> 
```

</details>


Output of `Resolve-AzError`

<details><summary>Click to view</summary>

```powershell
   HistoryId: 8


Message        : The type initializer for 'Microsoft.Azure.Commands.Common.AzModule' threw an exception.
StackTrace     :    at Microsoft.Azure.Commands.Common.RegisterAzModule.ProcessRecord()
Exception      : System.TypeInitializationException
InvocationInfo : {Register-AzModule}
Line           :   $VTable = Register-AzModule
                 
Position       : At C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\Az.MSGraph.psm1:49 char:13
                 +   $VTable = Register-AzModule
                 +             ~~~~~~~~~~~~~~~~~
HistoryId      : 8

Message        : Attempted to access an element as a type incompatible with the array.
StackTrace     :    at System.Collections.Generic.List`1.Add(T item)
                    at Microsoft.Azure.Commands.Common.AzModule..cctor()
Exception      : System.ArrayTypeMismatchException
InvocationInfo : {Register-AzModule}
Line           :   $VTable = Register-AzModule
                 
Position       : At C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\Az.MSGraph.psm1:49 char:13
                 +   $VTable = Register-AzModule
                 +             ~~~~~~~~~~~~~~~~~
HistoryId      : 8



   HistoryId: 5


Message        : The type initializer for 'Microsoft.Azure.Commands.Common.AzModule' threw an exception.
StackTrace     :    at Microsoft.Azure.Commands.Common.RegisterAzModule.ProcessRecord()
Exception      : System.TypeInitializationException
InvocationInfo : {Register-AzModule}
Line           :   $VTable = Register-AzModule
                 
Position       : At C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\Az.MSGraph.psm1:49 char:13
                 +   $VTable = Register-AzModule
                 +             ~~~~~~~~~~~~~~~~~
HistoryId      : 5

Message        : Attempted to access an element as a type incompatible with the array.
StackTrace     :    at System.Collections.Generic.List`1.Add(T item)
                    at Microsoft.Azure.Commands.Common.AzModule..cctor()
Exception      : System.ArrayTypeMismatchException
InvocationInfo : {Register-AzModule}
Line           :   $VTable = Register-AzModule
                 
Position       : At C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\Az.MSGraph.psm1:49 char:13
                 +   $VTable = Register-AzModule
                 +             ~~~~~~~~~~~~~~~~~
HistoryId      : 5



   HistoryId: 2


Message        : The type initializer for 'Microsoft.Azure.Commands.Common.AzModule' threw an exception.
StackTrace     :    at Microsoft.Azure.Commands.Common.RegisterAzModule.ProcessRecord()
Exception      : System.TypeInitializationException
InvocationInfo : {Register-AzModule}
Line           :   $VTable = Register-AzModule
                 
Position       : At C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\Az.MSGraph.psm1:49 char:13
                 +   $VTable = Register-AzModule
                 +             ~~~~~~~~~~~~~~~~~
HistoryId      : 2

Message        : Attempted to access an element as a type incompatible with the array.
StackTrace     :    at System.Collections.Generic.List`1.Add(T item)
                    at Microsoft.Azure.Commands.Common.AzModule..cctor()
Exception      : System.ArrayTypeMismatchException
InvocationInfo : {Register-AzModule}
Line           :   $VTable = Register-AzModule
                 
Position       : At C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\Az.MSGraph.psm1:49 char:13
                 +   $VTable = Register-AzModule
                 +             ~~~~~~~~~~~~~~~~~
HistoryId      : 2



   HistoryId: 1


Message        : The type initializer for 'Microsoft.Azure.Commands.Common.AzModule' threw an exception.
StackTrace     :    at Microsoft.Azure.Commands.Common.RegisterAzModule.ProcessRecord()
Exception      : System.TypeInitializationException
InvocationInfo : {Register-AzModule}
Line           :   $VTable = Register-AzModule
                 
Position       : At C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\Az.MSGraph.psm1:49 char:13
                 +   $VTable = Register-AzModule
                 +             ~~~~~~~~~~~~~~~~~
HistoryId      : 1

Message        : Attempted to access an element as a type incompatible with the array.
StackTrace     :    at System.Collections.Generic.List`1.Add(T item)
                    at Microsoft.Azure.Commands.Common.AzModule..cctor()
Exception      : System.ArrayTypeMismatchException
InvocationInfo : {Register-AzModule}
Line           :   $VTable = Register-AzModule
                 
Position       : At C:\Users\ORB\AppData\Local\Microsoft\PowerShell\Modules\Az.Resources\6.6.1\MSGraph.Autorest\Az.MSGraph.psm1:49 char:13
                 +   $VTable = Register-AzModule
                 +             ~~~~~~~~~~~~~~~~~
HistoryId      : 1
```

</details>

**Comment by o-l-a-v** [comment]
## Workaround for running locally
* Don't use VSCode + vscode-powershell + 5.1, or
* Install previous versions of mentioned modules

## Workaround for running in Azure Automation Account
* Download previous versions of mentioned modules manually from PowerShellGallery.
* Repackage them into a zip where
  * Zip file name: `<modulename>.zip`, example: `Az.Accounts.zip`
  * Root folder name: `<modulename>`, example: `Az.Accounts`
  * Child directory name: `<version>`, example: `2.12.1`.

```
Az.Accounts.zip
  Az.Accounts
    2.12.1
       <put content of nupkg file from PowerShellGallery here>
```

**Comment by bnichms** [comment]
> More info:
> 
> Running locally, `Az.Resources` v6.6.1 fails to import on Windows PowerShell 5.1, but only when using VSCode and the [VSCode PowerShell Extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.PowerShell).
> 
> * PowerShell.exe terminal works fine.
> * PowerShell ISE works fine
> * VSCode with PS extension had no problems with `Az.Accounts` v2.12.1 and `Az.Resources` v6.6.0.
> 
> Some assembly conflict?
> 
> Output with verbose and debug set to 'Continue'.
> 
> Click to view
> Output of `Resolve-AzError`
> 
> Click to view

Update: Issue is present in Powershell.exe (v5.1) on Windows Server

**Comment by isra-fel** [comment]
Thanks for the thorough investigation @o-l-a-v . We are looking into this problem.

**Comment by jrudley** [comment]
I am seeing this in GCC High (Azure Gov) as well.

**Comment by AlyaKoni** [comment]
Same here in several automation accounts. And not only Az.Resources. Also:
Az.Advisor
Az.Aks
Az.Compute
Az.DesktopVirtualization
Az.EventHub
Az.Kusto

**Comment by AlyaKoni** [comment]
In Runtime 7.1 already Az.Accounts failes:
Error message:
Error importing the module Az.Accounts. Import failed with the following error: Orchestrator.Shared.AsyncModuleImport.ModuleImportException: An error occurred when trying to import the module to an internal PowerShell session. Either the module dependencies are not imported correctly or the module is unsupported. Internal Error Message: Import-Module: C:\Users\Client\Temp\JGQKZQJGNR\Az.Accounts\Az.Accounts.psm1:87 Line | 87 | Import-Module (Join-Path -Path $PSScriptRoot -ChildPath Microsoft.Azu … | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ | Could not load file or assembly 'Newtonsoft.Json, Version=13.0.0.0, | Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed'. The system cannot | find the file specified. . at Orchestrator.Activities.SetModuleVersion.GetPs7ModuleVersion(String moduleName, String moduleFullPath, Guid accountId, Guid moduleVersionId, InitialSessionState defaultSessionState) at Orchestrator.Activities.SetModuleVersion.InsertModuleVersion(String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String moduleFullyQualifiedName, ActivitiesMetadataWorkflowExtension activitiesMetadataWorkflowExtension, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, ContentUriInfo contentUri, String storageUri, Int32 moduleVersion) at Orchestrator.Activities.SetModuleVersion.ExecuteInternal(CodeActivityContext context, String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String modulePath, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, String storageUri, Int32 moduleVersion) at Orchestrator.Activities.SetModuleVersion.Execute(CodeActivityContext context) at System.Activities.CodeActivity.InternalExecute(ActivityInstance instance, ActivityExecutor executor, BookmarkManager bookmarkManager) at System.Activities.Runtime.ActivityExecutor.ExecuteActivityWorkItem.ExecuteBody(ActivityExecutor executor, BookmarkManager bookmarkManager, Location resultLocation)

**Comment by MainDBA** [comment]
I have seen this issue too - same error as  above and runbooks that rely on these modules are failing/suspended...
The following updates in two of my Azure subscriptions (automated using an update runbook downloaded from MS's gallery) have failed on 30/05
AZ.Compute     from version 5.7.0 to 5.7.1
AZ.Resources   from version 6.6.0 to 6.6.1

Az.Accounts (which is a dependency of both) gets successfully updated to v 2.12.2.
Attempts to revert back to an earlier version of both modules also fail - when I delete the modules I am able to revert back to Az.Accounts v2.12.1 but then when I try and revert back to Az.Compute 5.7.0 or Az.Resources 6.6.0 they both first update Az.Accounts to v2.12.2 (despite documentation saying their dependency is v2.12.1) before installation and both attempts subsequently fail

**Comment by mortenlerudjordet** [comment]
Seeing the same problem after my AA account was auto updated to new Az version.
Multiple Az submodules are in a import failed state, and Runbooks trying to use them will just outright fail

**Comment by o-l-a-v** [comment]
Anyone tried Automation Account + _not_ Windows PowerShell 5.1? I haven't because newer PowerShell versions still are in preview (for Azure Automation Account).

Lets hope Microsoft can implement two new test cases that must succeed before launching a stable version of any `Az.*` module in the future:

* Azure Automation Account + Windows PowerShell 5.1
  * Import/install on the Automation Account itself
  * Import and use from a PowerShell runbook (easy to test auth with Managed Identity)
* VSCode + vscode-powershell + Windows PowerShell 5.1
  * Import

**Comment by tpascal** [comment]
Received the same exception while attempting to upgrade RunAs automation account from using AzureRM to Az

Manually installing Az.Accounts (2.12.1) and Az.Resources (6.6.0) via workaround gets the conversion completed and scripts running again with managed identity

**Comment by utkarshmsf** [comment]
**The Azure Automation team is investigating and working on the identified issue,** the workaround is to download the Az.Accounts v2.12.1 from the [PowerShell gallery ](https://www.powershellgallery.com/packages/Az.Accounts/2.12.1) and upload it with the following file structure: 

Az.Accounts.zip
|- Az.Accounts
  |- 2.12.1
    |- <put content of nupkg file from PowerShellGallery here>

All the modules having a dependency on the Az.Accounts module should work with the hot-fix.

**Comment by ritviknagpal48** [comment]
I hit a similar issue on one my environments: https://github.com/Azure/azure-powershell/issues/21690

**Comment by bnichms** [comment]
> I hit a similar issue on one my environments: #21690

@isra-fel please see this issue posted by one of our team members (which has relevant repro steps and other information)

**Comment by srjennings** [comment]
> > I hit a similar issue on one my environments: #21690
> 
> @isra-fel please see this issue posted by one of our team members (which has relevant repro steps and other information)


> FYI, I spoke with MS support on this and product is aware. They are working on a hotfix.

Hit this issue today. Do we have ETA on hotfix?

**Comment by jrudley** [comment]
https://learn.microsoft.com/en-us/azure/automation/automation-update-azure-modules

![image](https://user-images.githubusercontent.com/1934411/236318129-847904f5-eacf-4ab9-b33e-9b9627326fdd.png)

No eta yet per my Microsoft case I have opened.

**Comment by srjennings** [comment]
> https://learn.microsoft.com/en-us/azure/automation/automation-update-azure-modules
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/1934411/236318129-847904f5-eacf-4ab9-b33e-9b9627326fdd.png)
> 
> 
> 
> No eta yet per my Microsoft case I have opened.

Yeah. I had the same with Microsoft. Thanks for that doc link.

**Comment by shootex20** [comment]
Hi all, I've been encountering the same issue and have been trying to update the package manually due to the latest version of Az.Accounts causing me issues. My upload keeps failing even with the provided folder structure for reference. Currently, this is mine: 
![image](https://user-images.githubusercontent.com/19786531/236564623-b1687f33-bc49-48de-b336-f8c2d70adb0a.png)

Anyone have any ideas or able to help me out? Or will I have to wait for the update? 

This is the error I've got on my automation account 
![image](https://user-images.githubusercontent.com/19786531/236564891-ee9afaee-d210-4e3a-adf2-42332e935165.png)

Any information is appreciated. Thanks

**Comment by gvanbuskirkpps** [comment]
> Hi all, I've been encountering the same issue and have been trying to update the package manually due to the latest version of Az.Accounts causing me issues. My upload keeps failing even with the provided folder structure for reference. Currently, this is mine: ![image](https://user-images.githubusercontent.com/19786531/236564623-b1687f33-bc49-48de-b336-f8c2d70adb0a.png)
> 
> Anyone have any ideas or able to help me out? Or will I have to wait for the update?
> 
> This is the error I've got on my automation account ![image](https://user-images.githubusercontent.com/19786531/236564891-ee9afaee-d210-4e3a-adf2-42332e935165.png)
> 
> Any information is appreciated. Thanks

@shootex20 change the extension on the .nupkg file to be a .zip, extract the contents into that versioned folder, then try the import.

Az.Accounts.zip
| Az.Accounts
    | 2.12.1
        | --extracted contents of the nupkg archive

**Comment by isra-fel** [comment]
Update: root cause is that the way how the Azure Automation PS runtime is configured to resolve the reference to the Newtonsoft.Json library had an unexpected conflict with Az.Accounts, causing 2 versions of the assembly being loaded and used together, hence the error.
We are working with Automation team to modify the configuration to unblock Azure PowerShell.

**Comment by cblackuk** [comment]
We are hitting this issue on Azure Stack HCI 22H2 builds as well

```powershell
              Import-Module -Name "Az.StackHCI" -Force -ErrorAction Stop -Verbose
VERBOSE: Loading module from path 'C:\Program Files\WindowsPowerShell\Modules\Az.StackHCI\1.4.3\Az.StackHCI.psd1'.
VERBOSE: Removing the imported "Update-AzStackHciCluster" function.
VERBOSE: Removing the imported "Unregister-AzStackHCI" function.
VERBOSE: Removing the imported "Test-AzStackHCIConnection" function.
VERBOSE: Removing the imported "Set-AzStackHCI" function.
VERBOSE: Removing the imported "Remove-AzStackHCIVMAttestation" function.
VERBOSE: Removing the imported "Remove-AzStackHCIRemoteSupport" function.
VERBOSE: Removing the imported "Remove-AzStackHciExtension" function.
VERBOSE: Removing the imported "Remove-AzStackHciCluster" function.
VERBOSE: Removing the imported "Remove-AzStackHciArcSetting" function.
VERBOSE: Removing the imported "Register-AzStackHCI" function.
VERBOSE: Removing the imported "New-AzStackHciExtension" function.
VERBOSE: Removing the imported "New-AzStackHciCluster" function.
VERBOSE: Removing the imported "New-AzStackHciArcSetting" function.
VERBOSE: Removing the imported "Install-AzStackHCIRemoteSupport" function.
VERBOSE: Removing the imported "Get-AzStackHCIVMAttestation" function.
VERBOSE: Removing the imported "Get-AzStackHCIRemoteSupportSessionHistory" function.
VERBOSE: Removing the imported "Get-AzStackHCIRemoteSupportAccess" function.
VERBOSE: Removing the imported "Get-AzStackHciExtension" function.
VERBOSE: Removing the imported "Get-AzStackHciCluster" function.
VERBOSE: Removing the imported "Get-AzStackHciArcSetting" function.
VERBOSE: Removing the imported "Enable-AzStackHCIRemoteSupport" function.
VERBOSE: Removing the imported "Enable-AzStackHCIAttestation" function.
VERBOSE: Removing the imported "Disable-AzStackHCIRemoteSupport" function.
VERBOSE: Removing the imported "Disable-AzStackHCIAttestation" function.
VERBOSE: Removing the imported "Add-AzStackHCIVMAttestation" function.
VERBOSE: Cannot verify the Microsoft .NET Framework version 4.7.2 because it is not included in the list of permitted versions.
VERBOSE: Loading 'Assembly' from path 'C:\Program Files\WindowsPowerShell\Modules\Az.StackHCI\1.4.3\bin\Az.StackHCI.private.dll'.
VERBOSE: Loading 'Assembly' from path 'C:\Program Files\WindowsPowerShell\Modules\Az.StackHCI\1.4.3\bin\Az.StackHCI.private.dll'.
VERBOSE: Loading 'FormatsToProcess' from path 'C:\Program Files\WindowsPowerShell\Modules\Az.StackHCI\1.4.3\Az.StackHCI.format.ps1xml'.
VERBOSE: Populating RepositorySourceLocation property for module Az.StackHCI.
VERBOSE: Loading module from path 'C:\Program Files\WindowsPowerShell\Modules\Az.StackHCI\1.4.3\./Az.StackHCI.psm1'.
Register-AzModule : The type initializer for 'Microsoft.Azure.Commands.Common.AzModule' threw an exception.
At C:\Program Files\WindowsPowerShell\Modules\Az.StackHCI\1.4.3\Az.StackHCI.psm1:49 char:13
+   $VTable = Register-AzModule
+             ~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [Register-AzModule], TypeInitializationException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Common.RegisterAzModule

VERBOSE: Importing function 'Add-AzStackHCIVMAttestation'.
VERBOSE: Importing function 'Disable-AzStackHCIAttestation'.
VERBOSE: Importing function 'Disable-AzStackHCIRemoteSupport'.
VERBOSE: Importing function 'Enable-AzStackHCIAttestation'.
VERBOSE: Importing function 'Enable-AzStackHCIRemoteSupport'.
VERBOSE: Importing function 'Get-AzStackHciArcSetting'.
VERBOSE: Importing function 'Get-AzStackHciCluster'.
VERBOSE: Importing function 'Get-AzStackHciExtension'.
VERBOSE: Importing function 'Get-AzStackHCIRemoteSupportAccess'.
VERBOSE: Importing function 'Get-AzStackHCIRemoteSupportSessionHistory'.
VERBOSE: Importing function 'Get-AzStackHCIVMAttestation'.
VERBOSE: Importing function 'Install-AzStackHCIRemoteSupport'.
VERBOSE: Importing function 'New-AzStackHciArcSetting'.
VERBOSE: Importing function 'New-AzStackHciCluster'.
VERBOSE: Importing function 'New-AzStackHciExtension'.
VERBOSE: Importing function 'Register-AzStackHCI'.
VERBOSE: Importing function 'Remove-AzStackHciArcSetting'.
VERBOSE: Importing function 'Remove-AzStackHciCluster'.
VERBOSE: Importing function 'Remove-AzStackHciExtension'.
VERBOSE: Importing function 'Remove-AzStackHCIRemoteSupport'.
VERBOSE: Importing function 'Remove-AzStackHCIVMAttestation'.
VERBOSE: Importing function 'Set-AzStackHCI'.
VERBOSE: Importing function 'Test-AzStackHCIConnection'.
VERBOSE: Importing function 'Unregister-AzStackHCI'.
VERBOSE: Importing function 'Update-AzStackHciCluster'.
```

```powershell
 Get-Module -Name "Az.R*" -ListAvailable 


    Directory: C:\Program Files\WindowsPowerShell\Modules


ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     6.6.1      Az.Resources                        {Get-AzProviderOperation, Remove-AzRoleAssignment, Get-AzRoleAssignment, New-AzRoleAssignment...}

```

**Comment by skultetc** [comment]
Note for anyone still coming across this - Az.Sql 4.6.0 requires Az.Accounts 2.12.2, so if you're running SQL automations you'll need to install Az.Sql 4.5.0 manually as well.

**Comment by BartDecker** [comment]
Is this issue fixed in: az.Resources 6.7.0 and az.accounts 2.12.3?

**Comment by BartDecker** [comment]
> Not for vscode-powershell + Windows PowerShell 5.1 x64 at least. Got following error when trying to import Az.Monitor after having imported Az.Accounts v2.12.3.
> 
> ```powershell
> PS > Resolve-AzError
> Resolve-AzError : Method 'get_SerializationSettings' in type 'Microsoft.Azure.Management.Internal.Resources.ResourceManagementClient' from 
> assembly 'Microsoft.Azure.PowerShell.Clients.ResourceManager, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' does not 
> have an implementation.
> At line:1 char:1
> + Resolve-AzError
> + ~~~~~~~~~~~~~~~
>     + CategoryInfo          : NotSpecified: (:) [Resolve-AzError], TypeLoadException
>     + FullyQualifiedErrorId : System.TypeLoadException,Microsoft.Azure.Commands.Profile.Errors.ResolveError
> PS >
> ```
> 
> Tried to import Az.Resources with verbose enabled.
> 
> Click to expand
> Seems like a different error than previously.

Yeah, I also get that error now. At least something is changing :(

**Comment by jrudley** [comment]
Microsoft updated my case yesterday, but they had no progress update to report. This is for GCC High, but i'd imagine it applies to Commercial as well.

**Comment by o-l-a-v** [comment]
I created a dummy Automation Account to test with `Az.Accounts` v2.12.3.

`Az.Accounts` succeeded to 5.1 and 7.2 (but showing "unknown" module version), failed on 7.1.

![image](https://github.com/Azure/azure-powershell/assets/6450056/ef9d6ed3-b128-40fd-ba62-3bdf07ef3374)

Error for 7.1:

<details><summary>Click to expand</summary>

```
Error importing the module Az.Accounts. Import failed with the following error: Orchestrator.Shared.AsyncModuleImport.ModuleImportException: An error occurred when trying to import the module to an internal PowerShell session. Either the module dependencies are not imported correctly or the module is unsupported. Internal Error Message: Import-Module: C:\Users\Client\Temp\LZEBQTRPND\Az.Accounts\Az.Accounts.psm1:104 Line | 104 | Import-Module (Join-Path -Path $PSScriptRoot -ChildPath Microsoft.Azu … | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ | Could not load file or assembly 'Newtonsoft.Json, Version=13.0.0.0, | Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed'. The system cannot | find the file specified. . at Orchestrator.Activities.SetModuleVersion.GetPs7ModuleVersion(String moduleName, String moduleFullPath, Guid accountId, Guid moduleVersionId, InitialSessionState defaultSessionState) at Orchestrator.Activities.SetModuleVersion.InsertModuleVersion(String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String moduleFullyQualifiedName, ActivitiesMetadataWorkflowExtension activitiesMetadataWorkflowExtension, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, ContentUriInfo contentUri, String storageUri, Int32 moduleVersion) at Orchestrator.Activities.SetModuleVersion.ExecuteInternal(CodeActivityContext context, String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String modulePath, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, String storageUri, Int32 moduleVersion) at Orchestrator.Activities.SetModuleVersion.Execute(CodeActivityContext context) at System.Activities.CodeActivity.InternalExecute(ActivityInstance instance, ActivityExecutor executor, BookmarkManager bookmarkManager) at System.Activities.Runtime.ActivityExecutor.ExecuteActivityWorkItem.ExecuteBody(ActivityExecutor executor, BookmarkManager bookmarkManager, Location resultLocation)
```

</details>

Then tried to add latest `Az.Resource` to 5.1 and 7.2, as requirement of `Az.Accounts` v2.12.3 where not met on 7.1.

![image](https://github.com/Azure/azure-powershell/assets/6450056/646f462c-ae42-4dfc-9263-229f754900bd)

Succeeded on 7.2 but showing unknown version again. Failed of 5.1.

Error message for 5.1:

<details><summary>Click to expand</summary>

```
Error importing the module Az.Resources. Import failed with the following error: Orchestrator.Shared.AsyncModuleImport.ModuleImportException: An error occurred when trying to import the module to an internal PowerShell session. Either the module dependencies are not imported correctly or the module is unsupported. Internal Error Message: The type initializer for 'Microsoft.Azure.Commands.Common.AzModule' threw an exception. The type initializer for 'Microsoft.Azure.Commands.Common.AzModule' threw an exception. . at Orchestrator.Activities.SetModuleVersion.GetPsModuleVersion(String moduleName, String moduleFullPath, Guid accountId, Guid moduleVersionId, InitialSessionState defaultSessionState) at Orchestrator.Activities.SetModuleVersion.InsertModuleVersion(String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String moduleFullyQualifiedName, ActivitiesMetadataWorkflowExtension activitiesMetadataWorkflowExtension, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, ContentUriInfo contentUri, String storageUri, Int32 moduleVersion) at Orchestrator.Activities.SetModuleVersion.ExecuteInternal(CodeActivityContext context, String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String modulePath, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, String storageUri, Int32 moduleVersion) at Orchestrator.Activities.SetModuleVersion.Execute(CodeActivityContext context) at System.Activities.CodeActivity.InternalExecute(ActivityInstance instance, ActivityExecutor executor, BookmarkManager bookmarkManager) at System.Activities.Runtime.ActivityExecutor.ExecuteActivityWorkItem.ExecuteBody(ActivityExecutor executor, BookmarkManager bookmarkManager, Location resultLocation)
```

</details>

**Comment by BartDecker** [comment]
Got feedback that ETA for fix is 15-06

**Comment by o-l-a-v** [comment]
@BartDecker 

For what scenario? Automation Account vs. vscode-powershell + Windows PowerShell 5.1. Or both.

Who commits to fixing it? Az module maintainers vs. Automation Account maintainers vs. vscode-powershell.

**Comment by isra-fel** [comment]
There are a few things going on in this thread, different root causes, but all related to our recent update of `Newtonsoft.Json` library (the intention was to address security vulnerabilities which affected any version below 13). Allow me to summarize them:

### `Az.Resources` + Automation runbook + PS5

#### Symptom

Azure Automation runbook failed to import `Az.Resources` v6.6.1 for PowerShell 5.1 runtime.

#### Root cause

When importing `Az.Resources`, it executes the internal `Register-AzModule` cmdlet to connect it with `Az.Accounts` module for authentication and other purposes. The implementation of this cmdlet triggers the load of `NewtonSoft.Json`, a library for JSON serialization/deserialization. `Az.Accounts` did not expect the PowerShell runtime to provide versions lower than 13, so that the resolve can fall back to the assembly that ships with `Az.Accounts`. This is true in most cases, for example the PowerShell 5 that comes with Windows does not provide such an assembly. However, the PS runtime of Automation runbook **does** come with lower versions, causing part of `Az.Accounts` references v10, part references v13, and they can't work together.

#### Workarounds

[As suggested by @o-l-a-v , there are two workarounds](https://github.com/Azure/azure-powershell/issues/21647#issuecomment-1521820690) - downgrade Az modules or update PowerShell runtime to 7.2.

#### Solution

We have been working with Azure Automation team with this issue. Azure PowerShell implemented pre-loading in the latest version (2.12.3) of Az.Accounts, combined with the update of the runtime configuration of Azure Automation, we were able to verify in test environment that the issue was gone.

However, Azure Automation team needs more time to validate the solution in a larger scope. **We are targeting mid-June.**

Besides, both teams are working on adding integration tests to make sure the same thing doesn't happen again in the future.

### `Az.Resource` + VSCode PS extension + PS5

#### Symptom

The PS5 terminal provided by the PS VSCode extension cannot import `Az.Resources` v6.6.1.

#### Root cause

The PS5 runtime preloads `PowerShellEditorServices` module, which depends on `Newtonsoft.Json` 11.0. Similar to the previous issue, it caused conflict in Az.Accounts.

#### Workarounds

1. Use PowerShell 7 instead of 5 - go to the command palette of vscode, look for "PowerShell: Show Session Menu" and "Switch to: PowerShell (x64)".
2. Downgrade Az modules ([Az.Accounts v2.12.1](https://www.powershellgallery.com/packages/Az.Accounts/2.12.1), [Az.Resources v6.6.0](https://www.powershellgallery.com/packages/Az.Resources/6.6.0))
3. Use the PowerShell installed in your system.

#### Solution

We are reaching out to the team who maintains the `PowerShellEditorServices` for an update of the library. According to them, it is [inherited from the OmniSharp library that provides our LSP layer](https://github.com/PowerShell/PowerShellEditorServices/issues/2030#issuecomment-1570503787). We will then follow up with Omnisharp, which is part of .NET foundation.

### `Az.Accounts` + Automation runbook + PS7.1

#### Symptom

Azure Automation runbook failed to import `Az.Accounts` v2.12.2 for PowerShell 7.1 runtime.

#### Root cause

PowerShell 7.1 [reached end-of-support](https://learn.microsoft.com/en-us/lifecycle/products/powershell#:~:text=Nov%208%2C%202024-,PowerShell%207.1,-Nov%2011%2C%202020) last year, as well as [the support of PS7.1 from Azure PowerShell](https://learn.microsoft.com/en-us/powershell/azure/azureps-support-lifecycle?view=azps-9.7.1#supported-environments). It ships a lower version of `Newtonsoft.Json` than that `Az.Accounts` requires, which caused the issue.

#### Workarounds

[Downgrade Az modules or update PowerShell runtime to 7.2.](https://github.com/Azure/azure-powershell/issues/21647#issuecomment-1521820690).

**Comment by cblackuk** [comment]
> 

And what about Azure Stack HCI issue? :-)

**Comment by o-l-a-v** [comment]
@isra-fel 

We're passed mid-june now, could we get a status update? :)

**Comment by MainDBA** [comment]
My automatic PS Module updates have been on hold since late April and I needed to apply @o-l-a-v 's [workaround](https://github.com/Azure/azure-powershell/issues/21647#issuecomment-1521820690)  ... do we have any news on when this will be resolved please?

**Comment by dolauli** [comment]
Duplicate as https://github.com/Azure/azure-powershell/issues/21647.

**Comment by o-l-a-v** [comment]
> Duplicate as https://github.com/Azure/azure-powershell/issues/21960.

Eh, but this issue was created first?

**Comment by StevieLamb** [comment]
> 

I'm utterly shocked that these test cases are not in place already.
Given that Windows PowerShell is still
a) the only production-ready runtime in Azure Automation,
b) the required runtime for development of Runbooks on a Windows workstation, and
c) the default runtime targeted by MDM platforms

Without this testing, I can no longer recommend Azure Automation.
Now I need to look for an alternative on a competitor's infrastructure: critical code cannot be made to work on this platform without unstable, easy-to-break workarounds.

**Comment by StevieLamb** [comment]
> There are a few things going on in this thread, different root causes, but all related to our recent update of `Newtonsoft.Json` library (the intention was to address security vulnerabilities which affected any version below 13). Allow me to summarize them:
> 
> ### `Az.Resources` + Automation runbook + PS5
> #### Symptom
> Azure Automation runbook failed to import `Az.Resources` v6.6.1 for PowerShell 5.1 runtime.
> 
> #### Root cause
> When importing `Az.Resources`, it executes the internal `Register-AzModule` cmdlet to connect it with `Az.Accounts` module for authentication and other purposes. The implementation of this cmdlet triggers the load of `NewtonSoft.Json`, a library for JSON serialization/deserialization. `Az.Accounts` did not expect the PowerShell runtime to provide versions lower than 13, so that the resolve can fall back to the assembly that ships with `Az.Accounts`. This is true in most cases, for example the PowerShell 5 that comes with Windows does not provide such an assembly. However, the PS runtime of Automation runbook **does** come with lower versions, causing part of `Az.Accounts` references v10, part references v13, and they can't work together.
> 
> #### Workarounds
> [As suggested by @o-l-a-v , there are two workarounds](https://github.com/Azure/azure-powershell/issues/21647#issuecomment-1521820690) - downgrade Az modules or update PowerShell runtime to 7.2.
> 
> #### Solution
> We have been working with Azure Automation team with this issue. Azure PowerShell implemented pre-loading in the latest version (2.12.3) of Az.Accounts, combined with the update of the runtime configuration of Azure Automation, we were able to verify in test environment that the issue was gone.
> 
> However, Azure Automation team needs more time to validate the solution in a larger scope. **We are targeting mid-June.**
> 
> Besides, both teams are working on adding integration tests to make sure the same thing doesn't happen again in the future.
> 
> ### `Az.Resource` + VSCode PS extension + PS5
> #### Symptom
> The PS5 terminal provided by the PS VSCode extension cannot import `Az.Resources` v6.6.1.
> 
> #### Root cause
> The PS5 runtime preloads `PowerShellEditorServices` module, which depends on `Newtonsoft.Json` 11.0. Similar to the previous issue, it caused conflict in Az.Accounts.
> 
> #### Workarounds
> 1. Use PowerShell 7 instead of 5 - go to the command palette of vscode, look for "PowerShell: Show Session Menu" and "Switch to: PowerShell (x64)".
> 2. Downgrade Az modules ([Az.Accounts v2.12.1](https://www.powershellgallery.com/packages/Az.Accounts/2.12.1), [Az.Resources v6.6.0](https://www.powershellgallery.com/packages/Az.Resources/6.6.0))
> 3. Use the PowerShell installed in your system.
> 
> #### Solution
> We are reaching out to the team who maintains the `PowerShellEditorServices` for an update of the library. According to them, it is [inherited from the OmniSharp library that provides our LSP layer](https://github.com/PowerShell/PowerShellEditorServices/issues/2030#issuecomment-1570503787). We will then follow up with Omnisharp, which is part of .NET foundation.
> 
> ### `Az.Accounts` + Automation runbook + PS7.1
> #### Symptom
> Azure Automation runbook failed to import `Az.Accounts` v2.12.2 for PowerShell 7.1 runtime.
> 
> #### Root cause
> PowerShell 7.1 [reached end-of-support](https://learn.microsoft.com/en-us/lifecycle/products/powershell#:~:text=Nov%208%2C%202024-,PowerShell%207.1,-Nov%2011%2C%202020) last year, as well as [the support of PS7.1 from Azure PowerShell](https://learn.microsoft.com/en-us/powershell/azure/azureps-support-lifecycle?view=azps-9.7.1#supported-environments). It ships a lower version of `Newtonsoft.Json` than that `Az.Accounts` requires, which caused the issue.
> 
> #### Workarounds
> [Downgrade Az modules or update PowerShell runtime to 7.2.](https://github.com/Azure/azure-powershell/issues/21647#issuecomment-1521820690).

Are you genuinely advising customers to run production tasks using a runtime which is in preview on Azure Automation?
As I've stated elsewhere but to save redirection: switching to v7 is not viable for the above reason, nor in cases where we have no choice of the executing runtime - such as when deploying scripts via MS Intune.
Windows PowerShell is still the default runtime for Windows, regardless of how much everyone loves PowerShell Core.

**Comment by isra-fel** [comment]
Hello, I'm glad to announce that the integration issue between Azure PowerShell and Azure Automation runbook (on the 5.1 runtime) has been resolved and you are free to update the Azure PowerShell modules via the "Browse Gallery" feature.

At the meantime, we are discussing on bringing the later versions of Az to the "Update Az Modules" feature, which would save your effort updating all the Az modules.
What we are also working on is building the integration tests on both sides. We understand the missing of integration tests is dangerous and we are deeply sorry about the trouble it made.

**Comment by MasterKuat** [comment]
Hi @isra-fel,
Seems not, but take into account also case where automation account use a hybrid worker please.

**Comment by isra-fel** [comment]
> Hi @isra-fel, Seems not, but take into account also case where automation account use a hybrid worker please.

Thanks for the feedback @MasterKuat but could you please elaborate your scenario?

**Comment by MasterKuat** [comment]
I updated Az module on my hybrid worker yesterday. So Az.accounts was upgraded to 2.14.4.
Runbook fails : 
`Disable-AzContextAutosave : Method 'get_SerializationSettings' in type 'Microsoft.Azure.Management.Internal.Resources.ResourceManagementClient' from assembly 'Microsoft.Azure.PowerShell.Clients.ResourceManager, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' does not have an implementation. At C:\ProgramData\Microsoft\System Center\Orchestrator\7.2\SMA\Sandboxes\ue552ty2.5v2\Temp\o5uktgif.orn\rnb-grt-sysinfprd-we-BaseRunbook.ps1:47 char:13 + Disable-AzContextAutosave -Scope Process +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ + CategoryInfo : NotSpecified: (:) [Disable-AzContextAutosave], TypeLoadException + FullyQualifiedErrorId : System.TypeLoadException,Microsoft.Azure.Commands.Profile.Context.DisableAzureRmContextAutosave`

I downgraded Az.Accounts to 2.12.1 and it's ok.

By the way, we could not use PS7 as : It's a preview and PS7 runbook do not, as far as I know, permit to dot source other runbook.

**Comment by MrDRamos** [comment]
I'm having the same issues developing PowerShell 5.1 scripts within VsCode. Created a function to remove the incompatible module version based on a workaround from @o-l-a-v

```
function Repair-AzModulesForPS5()
{
    $ModuleS = @(
        @{Name = 'Az.Accounts';  MaxVer = '2.12.1'; Found = @(); Bad = @()}
        @{Name = 'Az.Resources'; MaxVer = '6.6.0';  Found = @(); Bad = @()}
        @{Name = 'Az.KeyVault';  MaxVer = '4.9.2';  Found = @(); Bad = @()}
    )

    foreach ($Module in $ModuleS) 
    {
        $Module.Found += Get-Module -Name $Module.Name -ListAvailable
        $Module.Bad += $Module.Found | Where-Object { $_.Version -gt $Module.MaxVer }
    }

    function Move_BadModules([array]$BadModuleS)
    {
        foreach ($Module in $BadModuleS) 
        {
            try 
            {
                $ScrDir = Split-Path -Path $Module.Path -Parent
                $DstDir = "$ScrDir`_incompatible"
                Remove-Item -Path $DstDir -Recurse -Force -ErrorAction Ignore
                Move-Item -Path $ScrDir -Destination $DstDir -Force -Verbose -ErrorAction Stop
            }
            catch 
            {
                Throw "Cant apply fix because $($Module.Name) $($Module.Version) was already loaded. Try again in new terminal session.`n$($_.Exception.Message)"
            }
        }
    }

    if ($ModuleS.Bad)
    {
        'VsCode + Powershell 5.1 +  Az-Modules are incompatible! (Thank you MS @#$%^)',
        'See: "https://github.com/Azure/azure-powershell/issues/21647#issuecomment-1521820690" ',
        'Will try to fix by replacing incompatible versions with older compatible ones:' | Write-Host -ForegroundColor Magenta
        
        foreach ($Module in $ModuleS) 
        {
            if ($Module.Found.Version -notcontains $Module.MaxVer)
            {
                Write-Host "Downloading compatible module: $($Module.Name) $($Module.MaxVer)"
                Install-Module -Name $Module.Name -RequiredVersion $Module.MaxVer          
            }
            Move_BadModules $Module.Bad
        }
    }
}    

```

**Comment by mbrat2005** [comment]
Running into this issue on Azure Stack HCI, which requires Windows PowerShell. The workaround that worked for us:

1. Run `Get-InstalledModule Az.Accounts,Az.Resources -AllVersion` 
2. Note Az.Accounts versions newer than `2.12.1` and Az.Resources newer than `6.6.0`
3. If not installed, install Az.Accounts 2.12.1 with `Install-Module Az.Accounts -requiredVersion 2.12.1` and do the same thing for Az.Resources version 6.6.0. 
4. Restart all PowerShell sessions (to unload any modules with dependencies) 
5. Run `Uninstall-Module Az.Accounts -RequiredVersion <newerVersion> -Force` for each version newer than 2.12.1
6. Run `Uninstall-Module Az.Resources -RequiredVersion <newerVersion> -Force` for each version newer than 6.6.0
7. Re-run `Get-InstalledModule Az.Accounts,Az.Resources -AllVersion` and ensure that the newest version of Az.Accounts is 2.12.1 and Az.Resources is 6.6.0

**Comment by insomniacc** [comment]
Seeing the same issues here when updating beyond 2.12.1 Az.Account, and Keyvault 4.9.2.
My situation is I'm using hybrid workers within an automation account, I'm using PS 5 since PS7 is still in preview and not really an option.

**Comment by AshishDadhich4h2** [comment]
What a mess :) Yeah it's breaking for me too.

**Comment by KezHalls** [comment]
Hi.  there is a lot of information to read here.  I have experienced the exact same issues with azure automation hybrid worker.  
Even though still in preview we opted to try the 7.2 version so I updated they hybrid worker server.
After running we get the below error.  Can I please get some guidance on peoples experience to next steps.

job...................................RunbookFlow : Method 'get_SerializationSettings' in type 'Microsoft.Azure.Management.Internal.Resources.ResourceManagementClient' from assembly 'Microsoft.Azure.PowerShell.Clients.ResourceManager, Version=1.0.0.0, Culture=neutral, PublicKeyToken=xxxx5' does not have an implementation. At line:1 char:57 + <#-- Enable activity tracing to see error location --#> RunbookFlow ` + ~~~~~~~~~~~~~ + CategoryInfo : NotSpecified: (:) [Invoke-RunbookFlow], TypeLoadException + FullyQualifiedErrorId : System.TypeLoadException,Orchestrator.GraphRunbook.Cmdlets.InvokeRunbookFlowCommand

**Comment by MasterKuat** [comment]
Hi @KezHalls , just to be sure, have you also recreate runbook to use runtime 7.1 ? Runtime version is fixed at runbook's creation time.

**Comment by mundayn** [comment]
Is there any update to this issue?

I am using a new Automation Account with Hybrid Worker Windows Server 2022 with Az modules freshly installed today using PowerShell 5.1 and experiencing the same issues.

Same in Visual Studio Code.

Receiving the "get_SerializationSettings" error, downgrading Az.Accounts and Az.Resources seemed to resolve it, until I needed to call another module using Az.Network which then needed a later version of Az.Accounts.

**Comment by brolifen** [comment]
> Is there any update to this issue?
> 
> I am using a new Automation Account with Hybrid Worker Windows Server 2022 with Az modules freshly installed today using PowerShell 5.1 and experiencing the same issues.
> 
> Same in Visual Studio Code.
> 
> Receiving the "get_SerializationSettings" error, downgrading Az.Accounts and Az.Resources seemed to resolve it, until I needed to call another module using Az.Network which then needed a later version of Az.Accounts.

Exactly the same. Just running Connect-AZAccount spits this error out on version 2.13.1. This is extremely annoying.

**Comment by brolifen** [comment]
Had to go down to 2.12.1 but first needed to manually remove the module from under C:\Program Files\WindowsPowerShell\Modules because it kept complaining it was still in use
Then ran:
Install-Module Az.Accounts -RequiredVersion 2.12.1 -Force

However as others mentioned this gives issues with other modules as some rely on the most recent one. Please fix this!

**Comment by o-l-a-v** [comment]
I've never worked with hybrid workers, but maybe manually installing and importing modules with a given version to a temp path on the workes could work?

Proof of concept script that worked from PowerShell ISE on Windows 10, and VSCode with PowerShell 5.1:

<details><summary>Click to expand</summary>
```powershell
#Requires -Version 5.1
#Requires -PSEdition Desktop
<#
    .SYNOPSIS
        Installs and imports Az modules to a given path. Adjust Assets to your requirements.

    .NOTES
        Author:   Olav Rønnestad Birkeland | github.com/o-l-a-v
        Created:  231008
        Modified: 231008

        # Why
        * 230425 - "Az.Resources v6.6.1 fails to import to Azure Automation Account #21647"
        https://github.com/Azure/azure-powershell/issues/21647

        # Learnings
        * Installing Az v9.6.0 with PowerShellGet or PackageManagement installs newest Az.Accounts.
          * Which means most modules won't load, because they rely on older versions of Az.Accounts.
        * Installing Az v9.6.0 with PowerShellGet or PackageManagement installs some other older Az submodules which rely on much older versions of Az.Accounts.
          * Az.PrivateDNS 1.0.3 for instance, requires <= Az.Accounts 1.8.1.

    .EXAMPLE
        & $(Try{$psEditor.GetEditorContext().CurrentFile.Path}Catch{$psISE.CurrentFile.FullPath})
#>


# Input parameters
[OutputType([System.Void])]
Param()


# Assets
$TempPath = [string] '{0}\Modules' -f $env:TEMP
$ModulesToInstall = [ordered]@{
    'Az.Accounts'  = [System.Version] '2.12.1'
    'Az.Compute'   = [System.Version] '5.7.0'
    'Az.Resources' = [System.Version] '6.6.0'
}


# Settings
$ErrorActionPreference = 'Stop'
$Global:ProgressPreference = 'SilentlyContinue'
if ([System.Net.ServicePointManager]::SecurityProtocol.ToString() -ne 'Tls12') {
    [System.Net.ServicePointManager]::SecurityProtocol = 'Tls12'
}


# Create path if it does not already exist
Write-Output -InputObject 'Create $TempPath if it does not already exist'
if (-not [System.IO.Directory]::Exists($TempPath)) {
    $null = [System.IO.Directory]::CreateDirectory($TempPath)
}


# Add info to $ModulesToInstall which will also be used for importing them later
$ModulesToInstall = [PSCustomObject[]](
    $ModulesToInstall.GetEnumerator().ForEach{
        [PSCustomObject]@{
            'Name'              = [string] $_.'Name'
            'Version'           = [System.Version] $_.'Value'
            'Uri'               = [string] 'https://www.powershellgallery.com/api/v2/package/{0}/{1}' -f $_.'Name', $_.'Value'
            'DownloadFilePath'  = [string] '{0}\{1}.{2}.zip' -f $TempPath, $_.'Name', $_.'Value'.ToString()
            'InstallDirectory'  = [string] '{0}\{1}\{2}' -f $TempPath, $_.'Name', $_.'Value'.ToString()
        }
    }.ForEach{
        $null = Add-Member -InputObject $_ -MemberType 'NoteProperty' -Name 'DetectionFilePath' -Value (
            [string] '{0}\{1}.psd1' -f $_.'InstallDirectory', $_.'Name'
        )
        $_
    }
)


# Install modules to $TempPath
Write-Output -InputObject ('{0}# Save modules from $ModulesToInstall to $TempPath' -f [System.Environment]::NewLine)
foreach ($Module in $ModulesToInstall) {
    Write-Output -InputObject ('{0} v{1}' -f $Module.'Name', $Module.'Version'.ToString())

    # Check if already installed
    if ([System.IO.File]::Exists($Module.'DetectionFilePath')) {
        Write-Output -InputObject 'Already installed.'
        Continue
    }

    # Download
    $null = [System.Net.WebClient]::new().DownloadFile($Module.'Uri',$Module.'DownloadFilePath')

    # Install
    ## Create directory
    $null = [System.IO.Directory]::CreateDirectory($Module.'InstallDirectory')
    ## Extract
    $null = Expand-Archive -Path $Module.'DownloadFilePath' -DestinationPath $Module.'InstallDirectory'
    ## Delete downloaded file
    $null = [System.IO.File]::Delete($Module.'DownloadFilePath')

    # Detect and output success
    if ([System.IO.File]::Exists($Module.'DetectionFilePath')) {
        Write-Output -InputObject 'Success.'
    }
    else {
        Write-Error -ErrorAction 'Stop' -Message 'Failed.' -Exception 'Failed.'
    }
}


# Import Az modules $TempPath
## Introduce step
Write-Output -InputObject ('{0}# Import all modules from $TempPath' -f [System.Environment]::NewLine)

# Remove any conflicting modules if already loaded
Write-Output -InputObject 'Remove any conflicting Az modules if already loaded.'
$null = Microsoft.PowerShell.Core\Get-Module -Name 'Az.*' | Remove-Module

## First Az.Accounts
Write-Output -InputObject 'Importing Az.Accounts'
Import-Module -Name $ModulesToInstall.Where{$_.'Name' -eq 'Az.Accounts'}.'DetectionFilePath' -WarningAction 'SilentlyContinue'

## Then other Az modules
Write-Output -InputObject ('Importing {0}' -f ($ModulesToInstall.Where{$_.'Name' -ne 'Az.Accounts'}.'Name' -join ', ' -as [string]))
Import-Module -Name $ModulesToInstall.Where{$_.'Name' -ne 'Az.Accounts'}.'DetectionFilePath'


# List imported modules
Write-Output -InputObject ('{0}# List imported modules' -f [System.Environment]::NewLine)
Get-Module | Select-Object -Property 'Name', 'Path' | ConvertTo-Json -Depth 1
```
</details>

**Comment by MasterKuat** [comment]
On a fresh HW (without any Az module) this should do the thing (not tested) : 

1. Install-Module Az.Accounts -RequiredVersion 2.12.1
1. Install-Module -Name Az -RequiredVersion 9.6.0 

Az 9.6.0 require minimum version of Az.Accounts 2.12.1 (https://www.powershellgallery.com/packages/Az/9.6.0/Content/Az.psm1)

Install-Module -Name Az always install lastest version of Az.Accounts even with a RequiredVersion (See #22335 )

**Comment by MrFly72** [comment]
But let's face it. That does not help you any if you need a newer version of another module which needs a higher version of AZ.Accounts. You are stuck in that moment. Don't know why there is so much talk about getting the old one. This problem needs a fix and it takes way too long. 
And as all modules have different requirements, it's a hazel to install them, as it is a huge work to walk through all other dependencies.

**Comment by brolifen** [comment]
> But let's face it. That does not help you any if you need a newer version of another module which needs a higher version of AZ.Accounts. You are stuck in that moment. Don't know why there is so much talk about getting the old one. This problem needs a fix and it takes way too long. And as all modules have different requirements, it's a hazel to install them, as it is a huge work to walk through all other dependencies.

Agreed installing older modules is just a workaround but I'm surprised this issue has been going on for so long for something so critical without a fix.

**Comment by MasterKuat** [comment]
MS ways past finding out but I guess they bet on PS 7.2

**Comment by brolifen** [comment]
> MS ways past finding out but I guess they bet on PS 7.2

Well at least on the Azure Automation side that is still labeled as "preview" aka not production ready but most importantly preview features are not under SLA or formal support. Not that the latter matters much with MS's support reputation as is evident by this open issue.

**Comment by mark3grahams** [comment]
https://github.com/Azure/azure-powershell/issues/21960

@isra-fel can you comment?

**Comment by o-l-a-v** [comment]
Just thought it was worth mentioning that PowerShell 7.2 is GA in Automation Account and works great with the latest Az modules.

* https://azure.microsoft.com/en-us/updates/azure-automation-powershell7-2-general-availability/

Seems 7.2 is GA for hybrid workers too?

* https://learn.microsoft.com/en-us/azure/automation/automation-hrw-run-runbooks
* https://learn.microsoft.com/en-us/azure/automation/automation-runbook-types

**Comment by MrFly72** [comment]
Any news on this?
It just stroke me again in VSCode:
![image](https://github.com/Azure/azure-powershell/assets/65505278/1fcd699a-fa6d-487c-8ef9-cfe5ea65c35c)
VSCode most recent, PS5.1, Powershell Extension most recent.
It is really a pain!

**Comment by Alex-wdy** [comment]
> Any news on this? It just stroke me again in VSCode: ![image](https://private-user-images.githubusercontent.com/65505278/294223912-1fcd699a-fa6d-487c-8ef9-cfe5ea65c35c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ0MTMxMjEsIm5iZiI6MTcwNDQxMjgyMSwicGF0aCI6Ii82NTUwNTI3OC8yOTQyMjM5MTItMWZjZDY5OWEtZmE2ZC00ODdjLThlZjktY2ZlNWVhNjVjMzVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA1VDAwMDAyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUwZGY2YmRiZWZkZjkwNmFhZDc2YzA4YjE1ZmQ3Y2I2MjU2NDliZmMxMDFhMDIxMTcyYTc4NWNjYmY3MjA3MzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Qpi27WDx7MbS6-LcapFpNJ8t1tn488bFn8hASUBN03g) VSCode most recent, PS5.1, Powershell Extension most recent. It is really a pain!

For VSCode Extension, we are tracking internally and we expect to solve this issue in January. I think all other issues have been resolved. Once I have a VsCode extension issue, I will update it in github.

**Comment by MasterKuat** [comment]
> Just thought it was worth mentioning that PowerShell 7.2 is GA in Automation Account and works great with the latest Az modules.

Yep but from my point of view 7.2 is NOT productionProof as : "Executing child scripts using .\child-runbook.ps1 is not supported in PowerShell 7.1 and PowerShell 7.2" (https://learn.microsoft.com/en-us/azure/automation/automation-child-runbooks#runbook-types) 
This is against the very first sentence on same page : "It's a good practice in Azure Automation to write reusable, modular runbooks with a discrete function that other runbooks call." 🤣

**Comment by MrFly72** [comment]
> For VSCode Extension issue, we are tracking internally, and we expect to solve this issue in January. I think all other issues have been resolved. Once I have a VsCode extension issue progress, I will update it in github.

I just tested the current preview (VSCode Extension ms-vscode.powershell@2024.1.0) and it still has this problem.
So is this fix you are talking about already in that?

**Comment by MrFly72** [comment]
> For VSCode Extension issue, we are tracking internally, and we expect to solve this issue in January. I think all other issues have been resolved. Once I have a VsCode extension issue progress, I will update it in github.

Can you give us an info here, when it is fixed in the preview of the powershell extension ?
As it is not easy to reproduce, i dont want to do it here and then.

**Comment by Alex-wdy** [comment]
Yes, I will continue to reply to this issue. A blog-issue postmortem should be posted at the same time.

**Comment by MrFly72** [comment]
@Alex-wdy I got a new prerelease (2024.3.0) of the extension and this is still not fixed in that.
This is really annoying. Please raise the priority of this, as we are all full stuck at a specific AZ release and if new features come in, we cannot use them.

**Comment by Alex-wdy** [comment]
> @Alex-wdy I got a new prerelease (2024.3.0) of the extension and this is still not fixed in that. This is really annoying. Please raise the priority of this, as we are all full stuck at a specific AZ release and if new features come in, we cannot use them.

This is on our high priority list, but will still take some time. I will update here as soon as there is any progress.

**Comment by ITJoeSchmo** [comment]
See my comment here for a potential workaround on PS 5.1: https://github.com/Azure/azure-powershell/issues/21960#issuecomment-1938831513

**Comment by MrFly72** [comment]
> > @Alex-wdy I got a new prerelease (2024.3.0) of the extension and this is still not fixed in that. This is really annoying. Please raise the priority of this, as we are all full stuck at a specific AZ release and if new features come in, we cannot use them.
> 
> This is on our high priority list, but will still take some time. I will update here as soon as there is any progress.

@Alex-wdy Still no news on this?
Tried everything with the newest extension + module, but still the same problem with 5.1?
I cannot believe that this blocking bug is open this long.
We are talking of one year now....
We just recently found out, that staying on the older modules had a function break in update-azvm, which resulted in making the azure Update Management break.
So there is a definite need for updated modules, which does not work if you have to develop in 5.1, as you have dependencies to modules that only run in 5.1 (in my case Citrix)

**Comment by ITJoeSchmo** [comment]
Same issue as here: https://github.com/Azure/azure-powershell/issues/21960 -- a conflict on the assembly NewtonSoft.Json due to modules loading 2 different versions and Windows PowerShell (aka 5.1 and older) do not know what to do when there are multiple versions of the same assembly loaded, causing this error. 

I posted a work-around and script to implement it in that same thread here: https://github.com/Azure/azure-powershell/issues/21960#issuecomment-2225896241

I have been leveraging this solution in 2 production automation platforms that use PowerShell in sandboxed processes (1 being AzureAutomation w/ HybridWorkers) without issue for nearly 6 months now. 

The only quirk is some modules have to be imported in a specific order to be happy if more than 1 is used in your script: 1) Az.* 2) Graph.* 3) ExchangeOnline 

I have only had to re-apply the bindingRedirect once when the app was updated and it was overwritten
